# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: layout/generic/nsFrameSelection.h
# Commit: e90eb44c2da3
# Full Hash: e90eb44c2da3303c038c67b7ad28b1498cd5a98c
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2024-07-19 16:21:39
# Description:
#   Bug 1897255 - Make `nsIFrame::PeekOffsetForLine` won't cross editing host boundary unless the callers allows that r=emilio
#   
#   The problems in the testcase are, `nsIFrame::PeekOffsetForLine` returns a frame
#   for any content node outside the editing host and `nsIFrame::GetLastLeaf`
#   returns a native anonymous subtree node if the frame is for the native
# ==============================================================================

diff -r bd02e7030721 -r e90eb44c2da3 layout/generic/nsFrameSelection.h
--- a/layout/generic/nsFrameSelection.h	Fri Jul 19 07:46:19 2024 +0000
+++ b/layout/generic/nsFrameSelection.h	Fri Jul 19 08:13:10 2024 +0000
@@ -112,7 +112,19 @@
                    // Passing by value here is intentional because EnumSet
                    // is optimized as uint*_t in opt builds.
                    const PeekOffsetOptions aOptions,
-                   EWordMovementType aWordMovementType = eDefaultBehavior);
+                   EWordMovementType aWordMovementType = eDefaultBehavior,
+                   const dom::Element* aAncestorLimiter = nullptr);
+
+  /**
+   * Return true if the ancestor limiter is not specified or if the content for
+   * aFrame is an inclusive descendant of mAncestorLimiter.
+   */
+  [[nodiscard]] bool FrameContentIsInAncestorLimiter(
+      const nsIFrame* aFrame) const {
+    return !mAncestorLimiter ||
+           (aFrame->GetContent() &&
+            aFrame->GetContent()->IsInclusiveDescendantOf(mAncestorLimiter));
+  }
 
   // Note: Most arguments (input and output) are only used with certain values
   // of mAmount. These values are indicated for each argument below.
@@ -156,6 +168,9 @@
 
   PeekOffsetOptions mOptions;
 
+  // The ancestor limiter element to peek offset.
+  const dom::Element* const mAncestorLimiter;
+
   /*** Output arguments ***/
 
   // Content reached as a result of the peek.